falcon/api.py

```diff
@@ -770,6 +764,13 @@ def _http_status_handler(self, req, resp, status, params):
     def _http_error_handler(self, req, resp, error, params):
         self._compose_error_response(req, resp, error)
 
+    def _find_error_handler(self, ex):
+        for exc in self._get_exception_lineage(ex):
```

Comment:
We should IMHO inline `type(ex).__mro__[:-1]` here for performance (maybe with a `PERF` note explaining why we are not reusing `get_exception_lineage`).
By saying for performance, I don't just mean avoiding an unnecessary function call which has become cheaper in CPython 3.7+, but also avoiding this check:
```python
        if not isinstance(ex, BaseException):
            raise TypeError('"exception" must be an exception type instance.')
```

It is important to check the type when adding handlers, but when handling an actual exception, we should presume it is of correct type :slightly_smiling_face: Even if in some unfathomable edge case it is not (old-style exceptions? although they were IIRC barred from Python 3), re-raising a `TypeError` would not be any more helpful.

Reply:
Sounds good. I pulled it out into a separate function for readability, but I think we can handle that fine with a `PERF` note as you suggested